### PR TITLE
Rework user identity handling and catch connectivity and login errors

### PR DESCRIPTION
**Pull request by jacoblurye**
This PR does a few things:
* creates the `AuthProvider` component for handling auth logic and providing auth data (e.g., id tokens) throughout the app via the `AuthContext`. Among other things, this re-implementation resolves warnings related to setting state in an existing state transition - now, auth data updates are proper state updates to `AuthProvider`'s state.
* creates the `UserProvider` component for handling account approval logic and providing user data (e.g., account info) through the app via `useUserContext`. This component separates out CIDC API calls from the Auth0 API calls that happen in `AuthProvider`.
* adds an `ErrorGuard` component for catching network connectivity and login errors.

`ErrorGuard` configured for network errors:
image

### CODE REVIEW

**Comment by curlup** [comment]
Couple of things

* Not sure that sending (changing url) user to `/error?...` is a way to go, because if they would want to provide us with a "faulting" link they will be unable to do so.
* I suggest to add `subject=...` to that " contact a CIDC Administrator" `mailto:...` link. Or even pre-populate body with the stuff we might want to know from such request - cimac account email, error date-time (to be able to filter logs) etc 
* I got this during login error (not sure if it's important): 
Screen Shot 2019-09-19 at 2 16 22 PM
* I see that after login error (I shut down the backend during login) it keeps sending those requests to `https://cidc-test.auth0.com/authorize` and `jwks.json` that end up with 200ok, but even with backend back online it never proceeds neither tries to connect to backend again.
* An attempt to open `/browse-files`while backend is off, never ends up with an error or something - it's still an infinite (well, as long as I can wait) spinner and nothing more.

**Comment by jacoblurye** [comment]
I believe https://github.com/CIMAC-CIDC/cidc-ui/pull/114/commits/7902655f766f7e8681cbfd86f744c45ae8ad8148 addresses the first and last bullet points.

I think inserting error info into the body of emails is a great idea, but I might suggest we leave that to a future PR (should be straightforward, though).

Re: that warning, I'm having some trouble reproducing, but given that it doesn't seem to be critical I'd say let's make an issue around it and save it for later.

File: .env [hunk 1]
```diff
@@ -6,4 +6,4 @@
 # Staging config
 REACT_APP_AUTH0_DOMAIN=cidc-test.auth0.com
 REACT_APP_AUTH0_CLIENT_ID=Yjlt8LT5vXFJw1Z8m8eaB5aZO26uPyeD
-REACT_APP_API_URL=https://staging-api.cimac-network.org
+REACT_APP_API_URL=http://localhost:5000
```
Comment (curlup) [inline .env]: =)

**Comment by curlup** [comment]
This one is still there I believe - if you open say `/` first, then shut down backend and try to click on "browse files"
> An attempt to open /browse-fileswhile backend is off, never ends up with an error or something - it's still an infinite (well, as long as I can wait) spinner and nothing more.

**Comment by curlup** [comment]
Same for `/user-account`

**Comment by jacoblurye** [comment]
Ah, gotcha - I misunderstood the error. Will craft a fix.

**Comment by jacoblurye** [comment]
Hmm, actually, I think a good fix to that issue would require an overhaul of the `api` module to set up error handling. Since the specific bug we were trying to fix this sprint was the confusing redirection-to-registration behavior when the API is already down before a user logs in, I suggest we delay these other fixes to another PR. Does that sound alright?

**Comment by curlup** [comment]
Totally fine with me!

**Pull request merged**